R/recipes.R

```diff
@@ -36,3 +36,59 @@ model_main_recipe <- function(data, pred_vars, cat_vars, id_vars) {
       strict = TRUE, zero_based = TRUE
     )
 }
+
+
+#' Create a Tidymodels recipe for linear baseline model
+#'
+#' @description Transforms categoricals using embeddings, normalizes numeric
+#'  predictors, creates interaction terms, and imputes missing values.
+#'
+#' @param data A data frame containing the input data.
+#' @param pred_vars Character vector of column names to keep for modeling. These
+#'   will be the right-hand side of the regression AKA predictors.
+#' @param cat_vars Character vector of categorical column names. These will be
+#'   transformed/encoded using embeddings.
+#' @param id_vars Character vector of ID variables. These can be kept in "baked"
+#'   data without being treated as predictors.
+#'
+#' @return A recipe object that can be used to clean model input data.
+#'
+model_lin_recipe <- function(data, pred_vars, cat_vars, id_vars) {
+  recipe(data) %>%
+    # Set the role of each variable in the input data
+    update_role(meta_sale_price, new_role = "outcome") %>%
+    update_role(all_of(pred_vars), new_role = "predictor") %>%
+    update_role(all_of(id_vars), new_role = "ID") %>%
+    update_role_requirements("ID", bake = FALSE) %>%
+    update_role_requirements("NA", bake = FALSE) %>%
+    # Remove any variables not an outcome var or in the pred_vars vector
+    step_rm(any_of("time_split")) %>%
+    step_rm(-all_outcomes(), -all_predictors(), -has_role("ID")) %>%
+    # Transforms and imputations
+    step_mutate(char_bldg_sf = ifelse(char_bldg_sf == 0, 1, char_bldg_sf)) %>%
+    step_mutate_at(starts_with("ind_"), fn = as.numeric) %>%
+    step_impute_median(all_numeric(), -all_outcomes(), -has_role("ID")) %>%
+    # Replace novel levels with "new"
+    step_novel(all_of(cat_vars), -has_role("ID")) %>%
+    # Replace NA in factors with "unknown"
+    step_unknown(all_of(cat_vars), -has_role("ID")) %>%
+    # Perform categorical embedding
+    embed::step_lencode_glm(
+      all_of(cat_vars), -has_role("ID"),
+      outcome = vars(meta_sale_price)
+    ) %>%
+    # Drop any predictors with near-zero variance, add interactions, and
+    # perform transforms
+    step_nzv(all_predictors()) %>%
+    step_interact(terms = ~ meta_township_code * time_sale_day) %>%
+    step_interact(terms = ~ meta_township_code * char_bldg_sf) %>%
+    step_BoxCox(
+      acs5_median_income_per_capita_past_year,
+      acs5_median_income_household_past_year,
+      char_bldg_sf
+    ) %>%
+    step_normalize(
```

Comment:
Only Q I have is: how did you decide which RHS columns to transform?

Reply:
Basically just a quick peek at the distributions of all the variables. These definitely aren't set in stone.